thrift/decode.go

```diff
@@ -0,0 +1,707 @@
+	d.r = r
+	d.f = d.f.without(protocolFlags).with(decoderFlags(r))
+}
+
+func (d *Decoder) SetStrict(enabled bool) {
+	if enabled {
+		d.f = d.f.with(strict)
+	} else {
+		d.f = d.f.without(strict)
+	}
+}
+
+func decoderFlags(r Reader) flags {
+	return flags(r.Protocol().Features() << featuresBitOffset)
+}
+
+var decoderCache atomic.Value // map[typeID]decodeFunc
+
+type decodeFunc func(Reader, reflect.Value, flags) error
+
+type decodeFuncCache map[reflect.Type]decodeFunc
+
+func decodeFuncOf(t reflect.Type, seen decodeFuncCache) decodeFunc {
+	f := seen[t]
+	if f != nil {
+		return f
+	}
+	switch t.Kind() {
+	case reflect.Bool:
+		f = decodeBool
+	case reflect.Int8:
+		f = decodeInt8
+	case reflect.Int16:
+		f = decodeInt16
+	case reflect.Int32:
+		f = decodeInt32
+	case reflect.Int64, reflect.Int:
+		f = decodeInt64
+	case reflect.Uint8:
+		f = decodeUint8
+	case reflect.Uint16:
+		f = decodeUint16
+	case reflect.Uint32:
+		f = decodeUint32
+	case reflect.Uint64, reflect.Uint, reflect.Uintptr:
+		f = decodeUint64
+	case reflect.Float32, reflect.Float64:
+		f = decodeFloat64
+	case reflect.String:
+		f = decodeString
+	case reflect.Slice:
+		if t.Elem().Kind() == reflect.Uint8 { // []byte
+			f = decodeBytes
+		} else {
+			f = decodeFuncSliceOf(t, seen)
+		}
+	case reflect.Map:
+		f = decodeFuncMapOf(t, seen)
+	case reflect.Struct:
+		f = decodeFuncStructOf(t, seen)
+	case reflect.Ptr:
+		f = decodeFuncPtrOf(t, seen)
+	default:
+		panic("type cannot be decoded in thrift: " + t.String())
+	}
+	seen[t] = f
+	return f
+}
+
+func decodeBool(r Reader, v reflect.Value, _ flags) error {
+	b, err := r.ReadBool()
+	if err != nil {
+		return err
+	}
+	v.SetBool(b)
+	return nil
+}
+
+func decodeInt8(r Reader, v reflect.Value, _ flags) error {
+	i, err := r.ReadInt8()
+	if err != nil {
+		return err
+	}
+	v.SetInt(int64(i))
+	return nil
+}
+
+func decodeInt16(r Reader, v reflect.Value, _ flags) error {
+	i, err := r.ReadInt16()
+	if err != nil {
+		return err
+	}
+	v.SetInt(int64(i))
+	return nil
+}
+
+func decodeInt32(r Reader, v reflect.Value, _ flags) error {
+	i, err := r.ReadInt32()
+	if err != nil {
+		return err
+	}
+	v.SetInt(int64(i))
+	return nil
+}
+
+func decodeInt64(r Reader, v reflect.Value, _ flags) error {
+	i, err := r.ReadInt64()
+	if err != nil {
+		return err
+	}
+	v.SetInt(int64(i))
+	return nil
+}
+
+func decodeUint8(r Reader, v reflect.Value, _ flags) error {
+	u, err := r.ReadInt8()
+	if err != nil {
+		return err
+	}
+	v.SetUint(uint64(u))
+	return nil
+}
+
+func decodeUint16(r Reader, v reflect.Value, _ flags) error {
+	u, err := r.ReadInt16()
+	if err != nil {
+		return err
+	}
+	v.SetUint(uint64(u))
+	return nil
+}
+
+func decodeUint32(r Reader, v reflect.Value, _ flags) error {
+	u, err := r.ReadInt32()
+	if err != nil {
+		return err
+	}
+	v.SetUint(uint64(u))
+	return nil
+}
+
+func decodeUint64(r Reader, v reflect.Value, _ flags) error {
+	u, err := r.ReadInt64()
+	if err != nil {
+		return err
+	}
+	v.SetUint(uint64(u))
```

Comment:
Should conversion issues be flagged (eg if the integer is <0)?

Why not returning a TypeMismatch?

Reply:
Hard to tell, since there is no unsigned integer type in thrift... it means we wouldn't be able to encode 64 bits unsigned integers where the most significant bit is 1.

Reply:
As a user I would not expect the library to support unsigned ints if the format doesn't. We could let the user do the castings and checks if they want to. Did you need it for parquet?

Reply:
Nope, we can drop all this if we don't want to allow unsigned integers 👍  I think I was mostly concerned about allowing what is allowed in `json` or `proto`, but it maybe is trying to be too smart.

Always a pleasure to remove code 🔪